zkvm/src/blockchain/state.rs

```diff
@@ -45,19 +45,83 @@ impl BlockchainState {
     fn apply_txlog(&mut self, txlog: &TxLog) -> Result<(), VMError> {
         for entry in txlog.iter() {
             match entry {
-                Entry::Input(contractID) => {
-                    // xxx check contractID is in self.utxos
-                    // xxx remove contractID
+                // Remove input from UTXO set
+                Entry::Input(input) => {
+                    let utxo = input.as_utxo();
+                    if self.utxos.contains(&utxo) {
+                        self.utxos.remove(&utxo);
+                    } else {
+                        return Err(VMError::InvalidInput);
+                    }
                 }
 
+                // Add output entry to UTXO set
                 Entry::Output(output) => {
-                    // xxx add output.id to self.utxos
+                    self.utxos.insert(UTXO::from_output(output));
                 }
-
                 _ => {}
             }
         }
 
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use curve25519_dalek::scalar::Scalar;
+    use rand::RngCore;
+
+    use super::*;
+    use crate::{Anchor, Contract, Data, Output, PortableItem, Predicate, VerificationKey};
+
+    fn rand_item() -> PortableItem {
+        let mut bytes = [0u8; 4];
+        rand::thread_rng().fill_bytes(&mut bytes);
+        PortableItem::Data(Data::Opaque(bytes.to_vec()))
+    }
+
+    fn rand_contract() -> Contract {
+        let mut bytes = [0u8; 32];
+        rand::thread_rng().fill_bytes(&mut bytes);
+        Contract {
+            anchor: Anchor::from_raw_bytes(bytes),
+            payload: vec![rand_item(), rand_item(), rand_item()],
+            predicate: Predicate::Key(VerificationKey::from_secret(&Scalar::random(
+                &mut rand::thread_rng(),
+            ))),
+        }
+    }
+
+    #[test]
+    fn test_apply_txlog() {
+        let mut state = BlockchainState::make_initial(0u64, 0u64);
+
+        // Add two outputs
+        let (output0, output1) = (Output::new(rand_contract()), Output::new(rand_contract()));
+        state
+            .apply_txlog(&vec![
+                Entry::Output(output0.clone()),
+                Entry::Output(output1.clone()),
+            ])
+            .unwrap();
+        state
+            .apply_txlog(&vec![Entry::Input(output0.id())])
+            .unwrap();
+
+        // Check that output0 was consumed
+        assert_eq!(state.utxos.contains(&output0.id().as_utxo()), false);
```

Comment:
Is there a reason to prefer this over `assert!(!state.utxos.contains(...))`?

Reply:
Not really! I'm happy to go with that

Reply:
`!(!`